features/device-config/monitor-adaptor/src/main/java/org/opennms/features/deviceconfig/monitor/adaptor/DeviceConfigMonitorAdaptor.java

```diff
@@ -151,11 +160,48 @@ private String getObjectAsString(Object object) {
         return null;
     }
 
+    private void cleanupStaleConfigs(Map<String, Object> parameters, OnmsIpInterface ipInterface, String serviceName) {
+        // Check for stale config records and delete them
+        final Date currentDate = new Date();
+        final String retentionPeriodPattern = getObjectAsString(parameters.get(DeviceConfigConstants.RETENTION_PERIOD));
+        final Period retentionPeriod = getRetentionPeriodOrDefault(retentionPeriodPattern);
+        final Date retentionDate = Date.from(currentDate.toInstant().atZone(ZoneId.systemDefault()).minus(retentionPeriod).toInstant());
+
+        // Get history of configs with config data, see if any exist that are not the latest one and are also
+        // older than the retention date
+        List<DeviceConfig> allConfigs = deviceConfigDao.findConfigsForInterfaceSortedByDate(ipInterface, serviceName);
+        Optional<DeviceConfig> lastGoodConfig = allConfigs.stream().filter(dc -> dc.getConfig() != null).findFirst();
```

Comment:
It may be more efficient to retrieve stale configs by SQL query itself.

Reply:
Good point, will take a look tomorrow.

Reply:
Updated this to use Dao